### PR TITLE
Consistently use linear indexing in copyto!(dest, src) - second try

### PR DESCRIPTION
**Pull request by pablosanjose**
I botched my first attempt (#34031) at fixing the inconsistency reported in #34025. The method responsible for the inconsistency was critical for the performance of `copyto!` between arrays with `IndexCartesian()` index style. My attempt at removing it in favor of the generic fallback in abstractarray.jl not only killed performance but even managed to break bootstrap somehow, I think 😮.

Here's my second attempt: the goal is to ensure that `copyto!(dest, src)` is  always equivalent to `for i in LinearIndices(src); dest[i] == src[i]; end`, without removing the mentioned fast path, but restricting it to the case of equal axes only, where both cartesian and linear indexing give the same result.

Some of the things that this fixes
```
julia> r = ones(Int, 3, 3); s = [1 2 3 4]'; copyto!(r, s)  # previously errored
3×3 Array{Int64,2}:
 1  4  1
 2  1  1
 3  1  1

julia> r = ones(Int, 3, 3); s = [2 0; 0 3]; copyto!(r, s) == copyto!(r, s')  # previously false
true

julia> r = ones(Int, 3, 3); s = [2 0; 0 3]; copyto!(r, s') # previously a cartesian copy, instead of linear copy
3×3 Array{Int64,2}:
 2  3  1
 0  1  1
 0  1  1

julia>  r = ones(Int, 3, 3); s = [2 0; 0 3]; @btime copyto!($r, $s)  # fastpath with the same performance as before
  6.036 ns (0 allocations: 0 bytes)

julia> r = ones(Int, 3, 3); s = [2 0; 0 3]; @btime copyto!($r, $(s')); # before was faster (16.1 ns) but incorrect
  22.390 ns (0 allocations: 0 bytes)
```

EDIT: I realize I should perhaps have simply updated the original PR, but after getting failed builds I thought I had messed something else badly, so I thought I should start fresh

### CODE REVIEW

**Comment by pablosanjose** [comment]
Could someone please restart the CI for me? I'm wondering if the test errors are spurious.

File: base/multidimensional.jl [hunk 1]
```diff
@@ -956,12 +956,22 @@ julia> y
...
+    if axes(dest) == axes(src)
+        for I in eachindex(IndexStyle(src′,dest), src′)
+            @inbounds dest[I] = src′[I]
+        end
+    # otherwise enforce linear indexing
+    else
+        linds = eachindex(IndexLinear(), src)
+        checkbounds(dest, first(linds))
+        checkbounds(dest, last(linds))
+        for I in linds
```
Comment (timholy) [inline base/multidimensional.jl]: The docs talk about "the first `n` elements" so I think this needs to ensure that we target the beginning of both arrays. Here's an example where the above fails:

```julia
a = fill(1, 5)
r = Base.IdentityUnitRange(-1:1)
copyto!(a, r)
```

**Review by timholy — COMMENTED** [review]
Sorry this hasn't gotten the attention it deserves. This looks good overall, save for one caveat I mention below in a comment. If this can be fixed and tested, then I think this should be merged at long last.

**Comment by pablosanjose** [comment]
Good catch @timholy, thanks! Fixed, hopefully.

**Comment by pablosanjose** [comment]
It seems we have broken tests by allowing behavior that previously errored. For example
```
a = OffsetArray{Int}(undef, (-3:-1,))
b = 1:2
@test_throws BoundsError copyto!(a, b)
```
This no longer throws. I feel uneasy about *removing* tests...

**Comment by pablosanjose** [comment]
Sorry for taking longer than I'd have wanted with this, I'm afraid I've been too busy lately.
I only needed to correct a couple of OffsetArrays tests that made assumptions incompatible with this PR.

**Comment by timholy** [comment]
Sorry I've been unresponsive myself. This looks great, many thanks @pablosanjose!

**Pull request merged**